### PR TITLE
TST: fixed cython doctests

### LINKED ISSUE #43757: TST: Add doctests on cython files
Adding `--doctest-cython` to `ci/run_tests.sh` to cover doctests in cython files.

Lots of them are failing at the moment. But I think few of them (like now, today), can be skipped using `# doctest: +SKIP`.

Below is the current list of failures atm.
```python
===================================================================================== short test summary info =====================================================================================
FAILED pandas/_libs/hashtable.cpython-38-darwin.so::pandas._libs.hashtable.Int64Factorizer.factorize
FAILED pandas/_libs/hashtable.cpython-38-darwin.so::pandas._libs.hashtable.ObjectFactorizer.factorize
FAILED pandas/_libs/hashtable.cpython-38-darwin.so::pandas._libs.hashtable.__test__.Int64Factorizer.factorize (line 134)
FAILED pandas/_libs/hashtable.cpython-38-darwin.so::pandas._libs.hashtable.__test__.ObjectFactorizer.factorize (line 88)
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.__test__.infer_dtype (line 1299)
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.__test__.is_iterator (line 242)
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.__test__.is_list_like (line 1051)
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.__test__.is_scalar (line 149)
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.infer_dtype
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.is_iterator
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.is_list_like
FAILED pandas/_libs/lib.cpython-38-darwin.so::pandas._libs.lib.is_scalar
FAILED pandas/_libs/tslibs/dtypes.cpython-38-darwin.so::pandas._libs.tslibs.dtypes.Resolution.from_attrname
FAILED pandas/_libs/tslibs/dtypes.cpython-38-darwin.so::pandas._libs.tslibs.dtypes.Resolution.get_reso_from_freq
FAILED pandas/_libs/tslibs/dtypes.cpython-38-darwin.so::pandas._libs.tslibs.dtypes.__test__.Resolution.from_attrname (line 223)
FAILED pandas/_libs/tslibs/dtypes.cpython-38-darwin.so::pandas._libs.tslibs.dtypes.__test__.Resolution.get_reso_from_freq (line 238)
FAILED pandas/_libs/tslibs/nattype.cpython-38-darwin.so::pandas._libs.tslibs.nattype.NaTType.now
FAILED pandas/_libs/tslibs/nattype.cpython-38-darwin.so::pandas._libs.tslibs.nattype.NaTType.today
FAILED pandas/_libs/tslibs/nattype.cpython-38-darwin.so::pandas._libs.tslibs.nattype.NaTType.utcfromtimestamp
FAILED pandas/_libs/tslibs/nattype.cpython-38-darwin.so::pandas._libs.tslibs.nattype.NaTType.utcnow
FAILED pandas/_libs/tslibs/period.cpython-38-darwin.so::pandas._libs.tslibs.period._Period.strftime
FAILED pandas/_libs/tslibs/period.cpython-38-darwin.so::pandas._libs.tslibs.period.__test__._Period.strftime (line 2243)
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.Timestamp.now
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.Timestamp.today
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.Timestamp.utcfromtimestamp
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.Timestamp.utcnow
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps._Timestamp.to_period
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.__test__.Timestamp.now (line 1048)
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.__test__.Timestamp.today (line 1075)
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.__test__.Timestamp.utcfromtimestamp (line 1115)
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.__test__.Timestamp.utcnow (line 1101)
FAILED pandas/_libs/tslibs/timestamps.cpython-38-darwin.so::pandas._libs.tslibs.timestamps.__test__._Timestamp.to_period (line 905)
FAILED pandas/_libs/tslibs/timezones.cpython-38-darwin.so::pandas._libs.tslibs.timezones.__test__.tz_standardize (line 375)
FAILED pandas/_libs/tslibs/timezones.cpython-38-darwin.so::pandas._libs.tslibs.timezones.tz_standardize
================================================================================= 34 failed, 106 passed in 1.10s ==================================================================================

### PR DESCRIPTION
**Pull request by debnathshoham**
- [x] closes #43757

Skipped some of the tests which were failing

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -60,6 +60,7 @@ dependencies:
   - flask
   - pytest>=6.0
   - pytest-cov
+  - pytest-cython
```
Comment (MarcoGorelli) [inline environment.yml:63]: might need to be in the `pip` section

Comment (debnathshoham) [reply environment.yml]: Ah, right!
amended.. thanks!

File: pandas/_libs/hashtable.pyx [hunk 2]
```diff
@@ -142,7 +142,7 @@ cdef class Int64Factorizer(Factorizer):
         --------
         Factorize values with nans replaced by na_sentinel
 
-        >>> factorize(np.array([1,2,np.nan], dtype='O'), na_sentinel=20)
+        >>> factorize(np.array([1,2,np.nan], dtype='O'), na_sentinel=20)# doctest: +SKIP
```
Comment (MarcoGorelli) [inline pandas/_libs/hashtable.pyx:145]: why does this one need skipping?

Comment (debnathshoham) [reply pandas/_libs/hashtable.pyx:145]: not sure why, it's not able to find `factorize` definition (also had similar issues in timezones.pyx)
would you have any idea?
```
_______________________________________________________________________ [doctest] pandas._libs.hashtable.Int64Factorizer.factorize _______________________________________________________________________
EXAMPLE LOCATION UNKNOWN, not showing all tests of that example
??? >>> factorize(np.array([1,2,np.nan], dtype='O'), na_sentinel=20)
UNEXPECTED EXCEPTION: NameError("name 'factorize' is not defined")
Traceback (most recent call last):
  File "/Users/debnathshoham/miniconda3/envs/pandas-dev/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest pandas._libs.hashtable.Int64Factorizer.factorize[0]>", line 1, in <module>
NameError: name 'factorize' is not defined
/Volumes/SD/pandas-debnathshoham/pandas/_libs/hashtable.cpython-38-darwin.so:None: UnexpectedException
______________________________________________________________________ [doctest] pandas._libs.hashtable.ObjectFactorizer.factorize _______________________________________________________________________
EXAMPLE LOCATION UNKNOWN, not showing all tests of that example
??? >>> factorize(np.array([1,2,np.nan], dtype='O'), na_sentinel=20)
UNEXPECTED EXCEPTION: NameError("name 'factorize' is not defined")
Traceback (most recent call last):
  File "/Users/debnathshoham/miniconda3/envs/pandas-dev/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest pandas._libs.hashtable.ObjectFactorizer.factorize[0]>", line 1, in <module>
NameError: name 'factorize' is not defined
/Volumes/SD/pandas-debnathshoham/pandas/_libs/hashtable.cpython-38-darwin.so:None: UnexpectedException

Comment (MarcoGorelli) [reply pandas/_libs/hashtable.pyx:145]: I don't know, perhaps you need the class name? (I haven't tried this)

File: pandas/_libs/tslibs/timestamps.pyx [hunk 3]
```diff
@@ -1087,7 +1091,7 @@ class Timestamp(_Timestamp):
 
         Examples
         --------
-        >>> pd.Timestamp.today()
+        >>> pd.Timestamp.today()    # doctest: +SKIP
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:1094]: why do these need skipping?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/hashtable.pyx (hunk 2 above)
Comment (debnathshoham) [reply pandas/_libs/hashtable.pyx]: Thanks! that did the trick in `ObjectFactorizer`.

Although for `IntFactorizer`, I am struggling to understand the use-case. The array can't contain `NaN` and be `int64`, am I missing something? (the original docstring was `object` dtype, which I think was incorrect)

Comment (MarcoGorelli) [reply pandas/_libs/hashtable.pyx]: I don't know, I've never used it - perhaps check `git log` and see if you can someone who's worked on it?

Comment (debnathshoham) [reply pandas/_libs/hashtable.pyx]: @jbrockmendel I see you have made some changes in `IntFactorizer`. Could you help me understand the use-case?

> The array can't contain NaN and be int64, am I missing something? (the original docstring was object dtype, which I think was incorrect)

File: pandas/_libs/tslibs/dtypes.pyx [hunk 4]
```diff
@@ -243,7 +243,7 @@ class Resolution(Enum):
 
         Examples
         --------
-        >>> Resolution.get_reso_from_freq('H')
+        >>> Resolution.get_reso_from_freq('H')  # doctest: +SKIP
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/dtypes.pyx:246]: why skipping these?

File: pandas/_libs/lib.pyx [hunk 5]
```diff
@@ -188,7 +189,7 @@ def is_scalar(val: object) -> bool:
     >>> pd.api.types.is_scalar((0, 2))
     False
 
-    pandas supports PEP 3141 numbers:
+    >>> # pandas supports PEP 3141 numbers:
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:192]: Can we avoid making this part of the doctest?

File: pandas/_libs/lib.pyx [hunk 6]
```diff
@@ -26,6 +26,7 @@ from cpython.tuple cimport (
     PyTuple_SET_ITEM,
 )
 from cython cimport floating
+import datetime
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:29]: i see why you're doing it, but im not wild about adding to the namespace just for doctests

Comment (MarcoGorelli) [reply pandas/_libs/lib.pyx:29]: I think you can add the import as part of the doctests?

Comment (debnathshoham) [reply pandas/_libs/lib.pyx]: yeah, done that now

File: pandas/_libs/lib.pyx (hunk 5 above)
Comment (debnathshoham) [reply pandas/_libs/lib.pyx]: skipped

File: pandas/_libs/tslibs/dtypes.pyx (hunk 4 above)
Comment (debnathshoham) [reply pandas/_libs/tslibs/dtypes.pyx]: the result was different. Changed that now.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Nice work, good to see this happening!

Couple of questions:
- In Python docstrings, you can add lines with descriptions between examples (e.g. in https://pandas.pydata.org/pandas-docs/stable/development/contributing_docstring.html#section-7-examples) - is that not possible for Cython docstrings?
- Is it possible to not skip the `tz_standardize` tests, for example by constructing `tz`?

**Comment by debnathshoham** [comment]
thanks @MarcoGorelli for reviewing this!

- yes I think that's valid for cython docstrings as well. I have removed the `# doctest: +SKIP` for the line `pandas supports PEP 3141 numbers:` and it's running fine.

- I have tried my best and was able to do the first two examples, but not able to replicate the third one (I don't have much experience working with `tz`).

So currently, only two tests are being skipped, i.e. last example in `tz` and `Int64Factorizer.factorize` in hashtable.pyx

File: pandas/_libs/hashtable.pyx [hunk 7]
```diff
@@ -142,7 +143,9 @@ cdef class Int64Factorizer(Factorizer):
         --------
         Factorize values with nans replaced by na_sentinel
 
-        >>> factorize(np.array([1,2,np.nan], dtype='O'), na_sentinel=20)
+        >>> fac = Int64Factorizer(3)
+        >>> arr = np.array([1,2,np.nan]).astype(np.int64)
+        >>> fac.factorize(arr, na_sentinel=20)    # doctest: +SKIP
```
Comment (jreback) [inline pandas/_libs/hashtable.pyx:148]: why is this skipped?

File: pandas/_libs/tslibs/timezones.pyx [hunk 8]
```diff
@@ -385,24 +385,26 @@ def tz_standardize(tz: tzinfo) -> tzinfo:
...
-
     >>> tz_standardize(tz)
     <DstTzInfo 'US/Pacific' LMT-1 day, 16:07:00 STD>
 
-    >>> tz
+    >>> tz        # doctest: +SKIP
     dateutil.tz.tz.tzutc
-
-    >>> tz_standardize(tz)
+    >>> tz_standardize(tz)    # doctest: +SKIP
```
Comment (jreback) [inline pandas/_libs/tslibs/timezones.pyx:407]: why does this need skipping?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/hashtable.pyx (hunk 7 above)
Comment (debnathshoham) [reply pandas/_libs/hashtable.pyx:148]: the original example was incorrect (`object` vs. `const int64_t`).
Tried to correct but I do not understand the use-case, how can array contain NaN and be int64?

File: pandas/_libs/tslibs/timezones.pyx (hunk 8 above)
Comment (debnathshoham) [reply pandas/_libs/tslibs/timezones.pyx:407]: I am not able to replicate this example

```python
In [33]: from dateutil.tz.tz import tzutc as tz

In [34]: tz
Out[34]: dateutil.tz.tz.tzutc

In [35]: from pandas._libs.tslibs.timezones import tz_standardize

In [36]: tz_standardize(tz)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-36-94d2ff882b92> in <module>
----> 1 tz_standardize(tz)

TypeError: Argument 'tz' has incorrect type (expected datetime.tzinfo, got _TzSingleton)

File: pandas/_libs/hashtable.pyx (hunk 7 above)
Comment (debnathshoham) [reply pandas/_libs/hashtable.pyx]: changed the example, removed NaN

File: pandas/_libs/tslibs/timezones.pyx [hunk 9]
```diff
@@ -385,24 +385,26 @@ def tz_standardize(tz: tzinfo) -> tzinfo:
     -------
     tzinfo
 
-    Examples:
+    Examples
     --------
+    >>> from datetime import datetime
+    >>> from pytz import timezone
+    >>> tz = timezone('US/Pacific').normalize(datetime(2014,1,1, \
+    tzinfo=pytz.utc)).tzinfo
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/timezones.pyx:393]: you can use `...` to split this over multiple lines, would prefer that to `\`. Something like:

```suggestion
    >>> tz = timezone('US/Pacific').normalize(
    ... datetime(2014, 1, 1, tzinfo=pytz.utc)
    ... ).tzinfo
```

**Comment by jbrockmendel** [comment]
> Edit - Another point worth highlighting, is the below is failing on my local (as the output is different than from the CI). [...] `??? >>> pd.Timestamp.fromtimestamp(1584199972)`

This is going to depend on what timezone you are in locally.

**Comment by debnathshoham** [comment]
> > Edit - Another point worth highlighting, is the below is failing on my local (as the output is different than from the CI). [...] `??? >>> pd.Timestamp.fromtimestamp(1584199972)`
> 
> This is going to depend on what timezone you are in locally.

would be a good idea to skip this no? (or maybe make it timezone independent if possible, to ensure it doesn't fail in anyones local)

File: pandas/_libs/tslibs/timezones.pyx (hunk 8 above)
Comment (MarcoGorelli) [reply pandas/_libs/tslibs/timezones.pyx:408]: @jreback sorry to unresolve this, but can we remove this example if it's not possible to run it?

**Comment by MarcoGorelli** [comment]
IMO the 
```
In [33]: from dateutil.tz.tz import tzutc as tz

In [34]: tz
Out[34]: dateutil.tz.tz.tzutc

In [35]: from pandas._libs.tslibs.timezones import tz_standardize

In [36]: tz_standardize(tz)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-36-94d2ff882b92> in <module>
----> 1 tz_standardize(tz)

TypeError: Argument 'tz' has incorrect type (expected datetime.tzinfo, got _TzSingleton)
```
test can be removed, if it doesn't run.

Other than that, the rest looks good to me

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @debnathshoham 

i suppose that if we had more doc-tests in cython this would pick them up. if someone wants to create an issue about adding more doc-strings wouldn't object (for example there is an older PR that adds a ton to the timestamp functions).

**Pull request merged**